components/tidb_query/src/rpn_expr/impl_cast.rs

```diff
@@ -323,6 +348,140 @@ fn cast_json_as_uint(ctx: &mut EvalContext, val: &Option<Json>) -> Result<Option
+                Ok(Some(Real::zero()))
+            } else {
+                // FIXME: TiDB here may has bug, fix this after fix TiDB's
+                Ok(Real::new(*val as u64 as f64).ok())
+            }
+        }
+    }
+}
+
+// because we needn't to consider if uint overflow upper boundary of signed real,
+// so we can merge uint to signed/unsigned real in one function
+#[rpn_fn]
+#[inline]
+fn cast_unsigned_int_as_signed_or_unsigned_real(val: &Option<Int>) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => Ok(Real::new(*val as u64 as f64).ok()),
+    }
+}
+
+#[rpn_fn]
+#[inline]
+fn cast_real_as_signed_real(val: &Option<Real>) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => Ok(Some(*val)),
+    }
+}
+
+#[rpn_fn(capture = [extra])]
+#[inline]
+fn cast_real_as_unsigned_real(
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Real>,
+) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            if in_union(extra.implicit_args) && val.into_inner() < 0f64 {
+                Ok(Some(Real::zero()))
+            } else {
+                Ok(Some(*val))
+            }
+        }
+    }
+}
+
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_string_as_signed_real(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra,
+    val: &Option<Bytes>,
+) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            // FIXME: in TiDB's builtinCastStringAsRealSig, if val is IsBinaryLiteral,
+            //  then return evalReal directly
+            let r: f64 = val.convert(ctx)?;
+            let r = produce_float_with_specified_tp(ctx, extra.ret_field_type, r)?;
+            Ok(Real::new(r).ok())
+        }
+    }
+}
+
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_string_as_unsigned_real(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra,
+    val: &Option<Bytes>,
+) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            // FIXME: in TiDB's builtinCastStringAsRealSig, if val is IsBinaryLiteral,
+            //  then return evalReal directly
+            let mut r: f64 = val.convert(ctx)?;
+            if in_union(extra.implicit_args) && r < 0f64 {
+                r = 0f64;
+            }
+            let r = produce_float_with_specified_tp(ctx, extra.ret_field_type, r)?;
+            Ok(Real::new(r).ok())
+        }
+    }
+}
+
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_decimal_as_unsigned_real(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Decimal>,
+) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            if in_union(extra.implicit_args) && val.is_negative() {
+                Ok(Some(Real::zero()))
+            } else {
+                // FIXME: here TiDB's may has bug, fix this after fix TiDB's
```

Comment:
ditto

Reply:
This is the same problem. Negative decimal to unsigned result may be wrong here.